### PR TITLE
Fix taint analysis for iterable block resolution

### LINKED ISSUE #9379: TaintAnalyzer fails with nested iterable operations.
Following program fails. 

```ballerina
import ballerina/io;

string pp = "street_number";
json tito;

function main(string... args) {
    j1.address_components
        .filter((json comp) => boolean {return comp.types.filter((json compType) => boolean {demo(); return compType.toString() == pp;}).count() > 0; })
        .foreach((json k) => { tito = k;});
    io:println(tito);
}


function demo() {
io:println("hola");
}

json j1 = {
        "address_components": [
            {
                "long_name": "1823",
                "short_name": "1823",
                "types": [
                    "street_number"
                ]
            },
            {
                "long_name": "CMW",
                "short_name": "CMW",
                "types": [
                    "postal_code_suffix"
                ]
            }
        ]
};
```

This results following error. 
```
ERROR {org.ballerinalang.launcher.Main} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.addToBlockedList(TaintAnalyzer.java:1716)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeLambdaExpressions(TaintAnalyzer.java:1678)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.lambda$visit$11(TaintAnalyzer.java:909)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:905)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:108)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:902)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:108)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:969)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangBinaryExpr.accept(BLangBinaryExpr.java:60)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:572)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangReturn.accept(BLangReturn.java:44)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1351)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:456)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:54)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1351)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:1574)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:1540)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visitInvokable(TaintAnalyzer.java:1507)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:328)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:66)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.lambda$visit$0(TaintAnalyzer.java:271)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:270)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:150)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyze(TaintAnalyzer.java:255)
	at org.wso2.ballerinalang.compiler.CompilerDriver.taintAnalyze(CompilerDriver.java:188)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:157)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:139)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:106)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:154)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:173)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:85)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:76)
	at org.ballerinalang.launcher.LauncherUtils.compile(LauncherUtils.java:257)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:87)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:256)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:66)
```

### PR DESCRIPTION
**Pull request by ayomawdb**
## Purpose
When performing block resolution for iterable operations, taint analyzer will try to fetch the symbol of the  "iterable operation invocation" to be stored as the blocking information used for debugging purposes. In iterable operations, this information (symbol) is not available leading to a NPE. The data gathered was simplified for iterable operations with this PR. 

Additional test cases were added to reflect the condition. 

Additionally, this PR will improve analysis of lambda functions written within iterable operations by analyzing the lambda function regardless of tainted status of the expression (collection). This is done to detect taint-errors within the lambda body.

Example: Calling a function that returns tainted data and passing that data to a security sensitive parameter, withing the lambda body). 

Unit tests are included for the improvement as well.

Resolves #9379

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 1]
```diff
@@ -1696,11 +1697,13 @@ private boolean processBlockedNode(BLangInvokableNode invokableNode, boolean isE
         return isBlocked;
     }
 
-    private void analyzeLambdaExpressions(BLangInvocation invocationExpr, BLangExpression argExpr) {
+    private void analyzeIterableLambdaInvocationExpression(BLangInvocation invocationExpr, BLangExpression argExpr) {
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Do we need to pass invocationExpr here?

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/taintchecking/TaintedStatusPropagationTest.java [hunk 2]
```diff
@@ -280,6 +280,22 @@ public void testIterableNegative() {
         BAssertUtil.validateError(result, 1, "tainted value passed to sensitive parameter 'secureIn'", 5, 53);
     }
 
+    @Test
+    public void testIterableWitinIterable() {
+        CompileResult result = BCompileUtil
+                .compile("test-src/taintchecking/propagation/iterable-within-iterable.bal");
```
Comment (anupama-pathirage) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/taintchecking/TaintedStatusPropagationTest.java]: Shall we reformat this to a single line?

File: tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable-negative.bal [hunk 3]
```diff
@@ -0,0 +1,38 @@
+json globalVar;
+
+function main (string... args) {
+    taintedJsonReturn().address_components
+    .filter((json comp) => boolean {return comp.types.filter((json compType) => boolean {secureFunction(taintedReturn(), taintedReturn()); return compType.toString() == "street_number";}).count() > 0; })
```
Comment (anupama-pathirage) [inline tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable-negative.bal]: Shall we reformat this to have 120 width?

File: tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable.bal [hunk 4]
```diff
@@ -0,0 +1,30 @@
+json globalVar;
+
+function main (string... args) {
+    jsonData.address_components
+    .filter((json comp) => boolean {return comp.types.filter((json compType) => boolean {secureFunction("untainted", "untainted"); return compType.toString() == "street_number";}).count() > 0; })
```
Comment (anupama-pathirage) [inline tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable.bal]: Shall we reformat this line and method signatures in this file

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/taintchecking/TaintedStatusPropagationTest.java (hunk 2 above)
Comment (ayomawdb) [reply tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/taintchecking/TaintedStatusPropagationTest.java]: +1. Will  also include other locations within this class.

File: tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable-negative.bal (hunk 3 above)
Comment (ayomawdb) [reply tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable-negative.bal]: +1

File: tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable.bal (hunk 4 above)
Comment (ayomawdb) [reply tests/ballerina-unit-test/src/test/resources/test-src/taintchecking/propagation/iterable-within-iterable.bal]: +1

**Comment by ayomawdb** [comment]
@anupama-pathirage I have made the changes relevant to the review comments.

**Pull request merged**